sass-spec.js

```diff
@@ -0,0 +1,192 @@
+const { archiveFromStream } = require("node-hrx")
+const tap = require("tap")
+const { promises: fs, createReadStream, rmdir } = require("fs")
+const path = require("path")
+const yaml = require("js-yaml")
+const child_process = require("child_process")
+
+/** Returns whether an options.yml object has a todo for the given impl */
+function hasTodo(options, impl) {
+  if (!options[":todo"]) return false
+  return options[":todo"].some((item) => item.includes(impl))
+}
+
+function hasIgnore(options, impl) {
+  if (!options[":ignore_for"]) return false
+  return options[":ignore_for"].includes(impl)
+}
+
+async function writeArchive(basePath, item) {
+  const fullPath = path.resolve(basePath, item.path)
+  if (item.isDirectory()) {
+    // If a directory, make the directory and recurse
+    await fs.mkdir(fullPath)
+    for (const subitem of item) {
+      await writeArchive(basePath, item.contents[subitem])
+    }
+  } else {
+    // We're a file, so write to it
+    await fs.writeFile(fullPath, item.body, { encoding: "utf-8" })
+  }
+}
+
+/**
+ * Unarchives the given HRX archive into the filesystem
+ */
+async function unarchive(parentDir, archiveName) {
+  // make a directory for the archive in the given directory
+  const dirName = archiveName.replace(".hrx", "")
+  const dirPath = path.resolve(parentDir, dirName)
+
+  // Unarchive and read the contents
+  const archivePath = path.resolve(parentDir, archiveName)
+  const archive = await archiveFromStream(
+    createReadStream(archivePath, { encoding: "utf-8" })
+  )
+  await writeArchive(dirPath, archive)
+}
+
+/**
+ * Run through the given directory
+ * @param dir the directory to iterate over
+ * @param opts the options to run the thing with:
+ *    - impl -- the sass implementation to use
+ *    - todo -- whether todos should be run
+ * @param cb the callback to run on each directory
+ */
+async function iterateDir(dir, opts, cb) {
+  // console.log(dir)
+  const { impl } = opts
+  const files = await fs.readdir(dir)
+  // If we find an options.yml file, read it and determine if we should go further
+  if (files.includes("options.yml")) {
+    const optsFile = path.resolve(dir, "options.yml")
+    const options = yaml.safeLoad(
+      await fs.readFile(optsFile, { encoding: "utf-8" })
+    )
+    // FIXME differentiate behavior of todos
+    // If the directory should be ignored or is a todo for this impl, do nothing else
+    if (hasTodo(options, impl) || hasIgnore(options, impl)) {
+      return
+    }
+  }
+  for (const filename of files) {
+    const filepath = path.resolve(dir, filename)
+    const filestat = await fs.stat(filepath)
+    if (filestat.isDirectory()) {
+      // If we run into a subdirectory, recurse into it
+      await iterateDir(filepath, opts, cb)
+    } else if (filename.endsWith(".hrx")) {
+      // If HRX, expand it into a directory and recurse into it
+      await unarchive(dir, filename)
```

Comment:
At some point, you'll probably want to avoid unarchiving until you actually need to run tests in a directory, and otherwise just using the HRX API to load file contents. Otherwise you'll be spending a lot of cycles on filesystem operations that you may not end up using, especially when running only a subset of specs.